### ISSUE TITLE
Generic.Aux compiles but errors with ClassCastException at runtime

### ISSUE THREAD

**Issue by matthughes** [milessabin/shapeless#1194]
Scala: 2.13.5
Shapeless: 2.3.7

I'm pretty positive that the following code used to not compile in Shapeless (note the private constructor and different return type of the companion apply):

```
case class Foo private(value: String)
object Foo {
    def apply(v: String): Either[String, Foo] = Left("No ways")
}
val gen: Generic.Aux[Foo,String :: HNil] = implicitly[Generic.Aux[Foo, String :: HNil]]
gen.from("test" :: HNil)
```

However, with the aforementioned versions, this does compile, but the last line will blow up at runtime with:

```"java.lang.ClassCastException: class scala.util.Left cannot be cast to class repl.MdocSession$App$Foo (scala.util.Left is in unnamed module of loader java.net.URLClassLoader @22fec1e1; repl.MdocSession$App$Foo is in unnamed module of loader 
scala.reflect.internal.util.AbstractFileClassLoader @43317cc1)
  at repl.MdocSession$App$$anonfun$2.apply(json.worksheet.sc:22)
  at repl.MdocSession$App$$anonfun$2.apply(json.worksheet.sc:22)
  at shapeless.Generic$$anon$1.from(generic.scala:164)
```

Note the above code will work if you make constructor public and remove the companion apply.

**Comment by milessabin** [comment]
I think this is most likely a consequence of [this change](https://github.com/scala/scala/commit/d6774a8bc7a0ee02b79a946f85f6dd344eb45b5f) which landed in Scala 2.13.6. As of that change there's no way to construct a value of your class `Foo`, so the example as written shouldn't compile.

**Comment by joroKr21** [comment]
I think this is a regression introduced by the optimization with helper methods to avoid anonymous classes. It looks like the previous implementation relied on type checking errors to catch problem like this but the current implementation uses an unsafe helper method (so it does casting). We should add a check that the return type matches in the code that looks for the `apply` method.

**Comment by joroKr21** [comment]
Fixed in #1197
